### PR TITLE
Support for CoNLL Corpora

### LINKED ISSUE #10: CoNLL support
It'd be good to support CoNLL format in a generic sense (and then perhaps some of the more specific CoNLL formats as an offshoot). I'd be happy to work on this if this is something you think would be worth it.

### PR DESCRIPTION
**Pull request by Ayushk4**
I am adding support for various CoNLL corpora in a generic sense and CoNLL 2003 in particular. ~I am assuming the deprecation fixes from #18.~

Can you review the data website used in the DataDeps registration block currently being used?

Refer #10.

### CODE REVIEW

File: src/CoNLL.jl [hunk 1]
```diff
@@ -0,0 +1,26 @@
+struct CoNLL{S}
+    filepaths::Vector{S}
+end
+
+function CoNLL(dirpath)
+    @assert(isdir(dirpath), dirpath)
+
+    files = []
+    for file in readdir(dirpath)
+        if length(file) > 4 && file[end-3:end] == ".txt"
```
Comment (oxinabox) [inline src/CoNLL.jl]: `endswith(file, ".txt")`
IIRC

File: src/CoNLL_DataDeps.jl [hunk 2]
```diff
@@ -0,0 +1,26 @@
+using DataDeps
+
+register(DataDep("CoNLL 2003",
```
Comment (oxinabox) [inline src/CoNLL_DataDeps.jl]: Can you check this does not conflict with the name MLDataSets.jl registers?

File: src/CoNLL_DataDeps.jl [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    The Seventh Conference on Natural Language Learning (CoNLL-2003) was held on May 31 and June 1, 2003 in association with HLT-NAACL 2003 in Edmonton, Canada.
+
+    The details of the task are in the following paper, please cite the following publication if you are using the corpora.
+
+    Erik F. Tjong Kim Sang, Fien De Meulder. "Introduction to the CoNLL-2003 Shared Task: Language-Independent Named Entity Recognition." Proceedings of CoNLL-2003, Edmonton, Canada, 2003.
+    https://www.clips.uantwerpen.be/conll2003/pdf/14247tjo.pdf
+    """,
+    "https://github.com/davidsbatista/NER-datasets/archive/master.zip",
+    "db332cddb12f123b92fc6653a582eec1b3f14e6741f9c6e0bf5d960c78dee3a0";
+    post_fetch_method = fn -> begin
```
Comment (oxinabox) [inline src/CoNLL_DataDeps.jl]:
```suggestion
    post_fetch_method = function(fn)
```

File: src/CoNLL.jl [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+            push!(files, file)
+        end
+    end
+
+    files = joinpath.(dirpath, files)
+    CoNLL(files)
```
Comment (oxinabox) [inline src/CoNLL.jl]:
```suggestion
    return CoNLL(files)
```

**Comment by oxinabox** [comment]
looks solid to me

**Comment by Ayushk4** [comment]
The CoNLL datasets have a separate test and train files. Should I modify the struct to hold the path of these files and then allow the user to call train or test files?
Somewhat like this- 
```
struct CoNLL{S}
    filepaths::Vector{S}
    year::Int
    trainpath::String
    testpath::String
    devpath::String
end
```
and it could be accessed something like following - `load(CoNLL(), "train")` which returns the files from train.txt.

**Comment by oxinabox** [comment]
Yes, that sopunds good.

File: src/types.jl [hunk 5]
```diff
@@ -68,16 +67,23 @@ struct SenseAnnotatedWord <: TaggedWord
...
 struct PosTaggedWord <: TaggedWord
     pos::String
     word::String
     function PosTaggedWord(pos, word)
-          new(intern(pos), intern(word))
+        new(intern(pos), intern(word))
+    end
+end
+
+struct CoNLL2003TaggedWord <: TaggedWord
```
Comment (oxinabox) [inline src/types.jl]: Can we just call this `NERTaggedWord` ?

File: src/CoNLL.jl [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+	tokens_tags = split(line)
+	length(tokens_tags) != 4 && throw("Error parsing line: \"$line\". Invalid Format.")
+	return CoNLL2003TaggedWord(tokens_tags[4], tokens_tags[3],
+	                           tokens_tags[2], tokens_tags[1])
+end
+
+function parse_conllfile(filename)
+    local sent
+    local doc
+    docs = @NestedVector(TaggedWord,3)()
```
Comment (oxinabox) [inline src/CoNLL.jl]: Don't make containers of Abstract types

File: src/CoNLL.jl [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    end
+    isempty(doc[end]) && deleteat!(doc, lastindex(doc))
+
+    return context
+end
+
+function load(corpus::CoNLL, file="train")
+    file == "train" && return parse_conllfile(corpus.trainpath)
+    file == "test" && return parse_conllfile(corpus.testpath)
+    file == "dev" && return parse_conllfile(corpus.devpath)
```
Comment (oxinabox) [inline src/CoNLL.jl]: Else error

**Pull request merged**